apps/pipelines/models.py

```diff
@@ -116,6 +124,26 @@ def invoke(self, input: PipelineState, session: ExperimentSession | None = None)
             pipeline_run.save()
         return output
 
+    def _create_pipeline_run(self, input: PipelineState, session: ExperimentSession) -> "PipelineRun":
+        # Django doesn't auto-serialize objects for JSON fields, so we need to copy the input and save the ID of
+        # the session instead of the session object.
+
+        return PipelineRun.objects.create(
+            pipeline=self,
+            input=input.json_safe(),
+            status=PipelineRunStatus.RUNNING,
+            log={"entries": []},
+            session=session,
+        )
+
+    def _save_message_to_history(self, session: ExperimentSession, message: str, type_: ChatMessageType):
+        ChatMessage.objects.create(
+            chat=session.chat,
+            message_type=type_.value,
+            content=message,
+        )
+        # TODO: Add tags here?
```

Comment:
We add tags for the router child bot (the "conditional" bot). I think it's fine to leave this as-is and only add tags here when we have support for branching

Reply:
I'm not sure that tagging makes sense for pipelines at all (unless we make it a config option for the nodes).

An alternate approach would be to link the chat messages to the pipeline run which is a bit like a trace.